x-pack/platform/plugins/shared/onechat/CONTRIBUTOR_GUIDE.md

```diff
@@ -0,0 +1,267 @@
+
+## Registering Built tools tools
+
+### Registering the tool
+
+Registering tools can be done using the `tools.register` API of the `onechat` plugin's setup contract.
+
+```ts
+class MyPlugin {
+  setup(core: CoreSetup, { onechat }: { onechat: OnechatPluginSetup }) {
+    onechat.tools.register(myToolDefinition);
+  }
+}
+```
+
+### Adding the tool to the allow list
+
+To allow the agent builder owners to control which tools are added to our framework, we are maintaining an hardcoded
+list of all internally registered tools. The intention is simply to trigger a code review from the team when
+tools are added, so that we can review it.
+
+To add a tool to the allow list, simply add the tool's id to the `AGENT_BUILDER_BUILTIN_TOOLS` array,
+in `x-pack/platform/packages/shared/onechat/onechat-server/allow_lists.ts`
+
+(Kibana will fail to start otherwise, with an explicit error message explaining what to do)
+
+### Registering built-in tools
+
+#### Basic example
+
+A simple example, with a tool just doing some math:
+
+```ts
+onechat.tools.register({
+  id: 'platform.examples.add_42',
+  type: ToolType.builtin,
+  description: 'Returns the sum of the input number and 42.',
+  tags: ['example'],
+  schema: z.object({
+    someNumber: z.number().describe('The number to add 42 to.'),
+  }),
+  handler: async ({ someNumber }) => {
+    return {
+      results: [
+        {
+          type: ToolResultType.other,
+          data: { value: 42 + someNumber },
+        },
+      ],
+    };
+  },
+});
+```
+
+#### using scoped services
+
+To let tools use services scoped to the current user during execution, we expose a set of services
+from the `context` object, exposed as the second parameter of the tool's handler.
+
+This context exposes, in addition to the `request` object, a panel of pre-scoped services such as:
+
+- scoped ES client
+- model provider (exposing a scoped inference client)
+- scoped logger
+
+```ts
+onechat.tools.register({
+  id: 'platform.examples.scoped_services',
+  type: ToolType.builtin,
+  description: 'Some example',
+  tags: ['example'],
+  schema: z.object({
+    indexPattern: z.string().describe('Index pattern to filter on'),
+  }),
+  handler: async ({ indexPattern }, { request, modelProvider, esClient }) => {
+    const indices = await esClient.asCurrentUser.cat.indices({ index: indexPattern });
+
+    const model = await modelProvider.getDefaultModel();
+    const response = await model.inferenceClient.chatComplete(somethingWith(indices));
+
+    const myCustomScopedService = await getMyCustomScopedService(request);
+    myCustomScopedService.doSomething(response);
+
+    return {
+      results: [{ type: ToolResultType.other, data: response }],
+    };
+  },
+});
+```
+
+Refer to `ToolHandlerContext` in `x-pack/platform/packages/shared/onechat/onechat-server/tools/handler.ts` to
+have access to the full list of services available from the handler context.
+
+#### reporting tool progress
```

Comment:
suggestion: add steps on how to customise the tool UI rendering in the thinking panel? Fine with deferring this into a separate ticket until theres a real ask for this as will likely require some additional changes in FE code to make it easier to do this.

Reply:
Well, atm there's no way/API for doing that, is there 🤔 ?